### PR TITLE
[ALLUXIO-2830]Shade Swift, GlusterFs and HDFS

### PR DESCRIPTION
**Pull request by apc999**
https://alluxio.atlassian.net/browse/ALLUXIO-2830

### CODE REVIEW

**Comment by apc999** [comment]
@calvinjia @jsimsa Please review the PR, with focus on the pom change. thanks

File: underfs/glusterfs/pom.xml [hunk 1]
```diff
@@ -50,10 +62,332 @@
...
+            <goals>
+              <goal>shade</goal>
+            </goals>
+            <configuration>
+              <transformers>
+                <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"/>
+              </transformers>
+              <artifactSet>
+                <includes>
+                  <include>org.apache.hadoop:hadoop-*</include>
```
Comment (jsimsa) [inline underfs/glusterfs/pom.xml]: do we need to shade guava as well?

File: underfs/hdfs/pom.xml [hunk 2]
```diff
@@ -64,4 +63,342 @@
...
+            <goals>
+              <goal>shade</goal>
+            </goals>
+            <configuration>
+              <transformers>
+                <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"/>
+              </transformers>
+              <artifactSet>
+                <includes>
+                  <include>org.apache.hadoop:hadoop-*</include>
```
Comment (jsimsa) [inline underfs/hdfs/pom.xml]: what about other external dependencies (guava and commons-lang3)?

Comment (jsimsa) [inline underfs/hdfs/pom.xml]: also, is this the superset of all dependencies of hadoop-2.x across different hadoop versions?

File: underfs/swift/pom.xml [hunk 3]
```diff
@@ -55,4 +53,77 @@
...
+          <execution>
+            <id>shade-swift</id>
+            <phase>package</phase>
+            <goals>
+              <goal>shade</goal>
+            </goals>
+            <configuration>
+              <artifactSet>
+                <includes>
+                  <include>org.codehaus.jackson:jackson-mapper-asl</include>
```
Comment (jsimsa) [inline underfs/swift/pom.xml]: what about guava?

**Comment by jsimsa** [comment]
@apc999 thanks for the meticulous work ... did a pass and left some comments

File: underfs/swift/pom.xml (hunk 3 above)
Comment (apc999) [inline underfs/swift/pom.xml]: talked offline. we are fine to leave guava outside the output jar

File: underfs/glusterfs/pom.xml (hunk 1 above)
Comment (apc999) [inline underfs/glusterfs/pom.xml]: talked offline. we are fine to leave guava outside the output jar

File: core/common/src/main/java/alluxio/underfs/UnderFileSystem.java [hunk 4]
```diff
@@ -97,7 +97,7 @@ public static UnderFileSystem create(String path, Map<String, String> ufsConf) {
               factory.create(path, new UnderFileSystemConfiguration(ufsConf)));
         } catch (Exception e) {
           errors.add(e);
-          LOG.warn("Failed to create ufs", e);
+          LOG.warn("Failed to create UnderFileSystemFactory {}", factory, e);
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystem.java]: Could you update this to just log the exception message?

File: pom.xml [hunk 5]
```diff
@@ -805,7 +811,8 @@
         <plugin>
           <groupId>org.apache.maven.plugins</groupId>
           <artifactId>maven-shade-plugin</artifactId>
-          <version>2.2</version>
+          <!-- Needs to be 2.4.2 or later to prevent MSHADE-148 -->
+          <version>2.4.3</version>
```
Comment (calvinjia) [inline pom.xml]: Any other side effects of upgrading this?

File: underfs/hdfs/pom.xml (hunk 2 above)
Comment (calvinjia) [inline underfs/hdfs/pom.xml]: Why do we use `:hadoop-*` instead of `:*`?

File: pom.xml (hunk 5 above)
Comment (apc999) [inline pom.xml]: I don't really find any side effect so far. 
Without this upgrade, the build fails non-deterministically.

File: pom.xml [hunk 6]
```diff
@@ -388,6 +388,51 @@
       </dependency>
       <dependency>
         <groupId>org.apache.hadoop</groupId>
+        <artifactId>hadoop-common</artifactId>
```
Comment (apc999) [inline pom.xml]: we still need these as they are introduced from north bound (e.g., hdfs-client, keyvalue-client) into assembly jar, even after this PR hadoop dependency is shaded from south bound

**Comment by apc999** [comment]
@jsimsa @calvinjia  PTAL

File: pom.xml (hunk 5 above)
Comment (apc999) [inline pom.xml]: note that, this is not really user facing but only related developers

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by apc999** [comment]
Jenkins, test this please

File: underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java [hunk 7]
```diff
@@ -87,12 +88,15 @@ public static HdfsUnderFileSystem createInstance(
    * @param conf the configuration for this UFS
    * @param hdfsConf the configuration for HDFS
    */
-  protected HdfsUnderFileSystem(AlluxioURI ufsUri, UnderFileSystemConfiguration conf,
+  public HdfsUnderFileSystem(AlluxioURI ufsUri, UnderFileSystemConfiguration conf,
```
Comment (jsimsa) [inline underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java]: optional: could this be package private?

Comment (apc999) [inline underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java]: done

**Comment by jsimsa** [comment]
Jenkins, test this please

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by calvinjia** [comment]
Jenkins, test this please.

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by calvinjia** [comment]
Jenkins, test this please.

**Comment by apc999** [comment]
Jenkins, test this please.

**Pull request merged**